### PR TITLE
Refactored Convolve to work with focal infrastructure.

### PR DESCRIPTION
**Pull request by lossyrob**

### CODE REVIEW

**Comment by lossyrob** [comment]
This is still WIP, because I need to do some benchmarks to ensure that I haven't slowed down the focal infrastructure.

**Comment by lossyrob** [comment]
Finished benchmarking, achieving speeds very close to 0.9 for rasters with lots of NODATA values, beating handedly for rasters containing no nodata

File: raster/src/main/scala/geotrellis/raster/op/focal/Kernel.scala [hunk 1]
```diff
@@ -25,9 +9,15 @@ import geotrellis.vector.Extent
  * Represents a neighborhood that is represented by
  * a tile.
  */
-case class Kernel(tile: Tile) {
+case class Kernel(tile: Tile) extends Neighborhood {
   if(tile.rows != tile.cols) sys.error("Kernel tile must be square")
```
Comment (echeipesh) [inline raster/src/main/scala/geotrellis/raster/op/focal/Kernel.scala]: Should this be `require`? It produces `IllegalArgumentException`, subtype of `RuntimeException` produced here.

File: spark/src/main/scala/geotrellis/spark/op/focal/FocalOperation.scala [hunk 2]
```diff
@@ -17,8 +17,9 @@ import scala.collection.mutable.ArrayBuffer
 
 object FocalOperation {
 
+  // TODO: Remove Neighborhood out of calc.
```
Comment (echeipesh) [inline spark/src/main/scala/geotrellis/spark/op/focal/FocalOperation.scala]: TODO done :)

File: raster/src/main/scala/geotrellis/raster/op/focal/Convolve.scala [hunk 3]
```diff
@@ -0,0 +1,89 @@
+package geotrellis.raster.op.focal
+
+import geotrellis.raster._
+import spire.syntax.cfor._
+
+/**
+ * Computes the convolution of a raster with a kernel.
+ *
+ * @param      r       Tile to convolve.
```
Comment (echeipesh) [inline raster/src/main/scala/geotrellis/raster/op/focal/Convolve.scala]: orphaned docs

File: raster/src/main/scala/geotrellis/raster/op/focal/FocalCalculation.scala [hunk 4]
```diff
@@ -44,17 +41,35 @@ abstract class FocalCalculation[T](
...
+  def calc(tile: Tile, cursor: Cursor): Unit
+}
+
+/**
+ * A focal calculation that uses the Cursor focal strategy.
+ */
+abstract class KernelCalculation[T](tile: Tile, kernel: Kernel, val analysisArea: Option[GridBounds]) 
+    extends FocalCalculation[T](tile, kernel, analysisArea)
+{
+  def traversalStrategy = ScanLineTraversalStrategy//TraversalStrategy.DEFAULT
```
Comment (echeipesh) [inline raster/src/main/scala/geotrellis/raster/op/focal/FocalCalculation.scala]: Out of curiosity why is `ScanLineTraversalStrategy` faster for convolve ?

**Comment by echeipesh** [comment]
+1 after my nitpicks.

**Pull request merged**